yeat/config/auto_pop.py

```diff
@@ -0,0 +1,109 @@
+#
+# This software was prepared for the Department of Homeland Security (DHS) by the Battelle National
+# Biodefense Institute, LLC (BNBI) as part of contract HSHQDC-15-C-00064 to manage and operate the
+# National Biodefense Analysis and Countermeasures Center (NBACC), a Federally Funded Research and
+# Development Center.
+# -------------------------------------------------------------------------------------------------
+
+from collections import defaultdict
+import json
+from pathlib import Path
+
+
+EXTENSIONS = (".fastq", ".fastq.gz", ".fq", ".fq.gz")
+
+
+class AutoPopError(ValueError):
+    pass
+
+
+class AutoPop:
+    def __init__(self, samples, seq_path, files):
+        self.samples = self.get_samples(samples)
+        self.check_samples()
+        self.check_seq_path(seq_path)
+        self.files = [Path(f) for f in files] if files is not None else self.get_files(seq_path)
+        self.check_files()
+        self.files_to_samples = self.organize_files_to_samples()
+        self.check_files_to_samples()
+
+    def get_samples(self, samples):
+        if len(samples) == 1 and Path(samples[0]).is_file():
+            with open(samples[0], "r") as fh:
+                samples = fh.read().strip().split("\n")
+        return list(sorted(samples))
+
+    def check_samples(self):
+        for s1_index, s1 in enumerate(self.samples):
+            for s2_index, s2 in enumerate(self.samples):
+                if s1 == s2 and s1_index == s2_index:
+                    continue
+                if s1 in s2:
+                    message = f"cannot correctly process a sample name that is a substring of another sample name: {s1} vs. {s2}"
+                    raise AutoPopError(message)
+
+    def check_seq_path(self, seq_path):
+        if not seq_path:
+            return
+        if not Path(seq_path).exists():
+            message = f"path does not exist: '{seq_path}'"
+            raise AutoPopError(message)
+
+    def get_files(self, seq_path):
+        files = []
+        for file in self.traverse(seq_path):
+            if not file.name.endswith(EXTENSIONS):
+                continue
+            files.append(file)
+        return files
+
+    def traverse(self, dirpath):
+        dirpath = Path(dirpath)
+        if not dirpath.is_dir():
+            return  # pragma: no cover
+        for subpath in dirpath.iterdir():
+            if subpath.is_dir():
+                yield from self.traverse(subpath)
+            else:
+                yield subpath
+
+    def check_files(self):
+        for file in self.files:
+            if not file.exists():
+                message = f"file does not exist: {file}"
+                raise AutoPopError(message)
+
+    def organize_files_to_samples(self):
+        files_to_samples = defaultdict(list)
+        for sample in self.samples:
+            files_to_samples[sample] = [file for file in self.files if sample in str(file)]
+        return files_to_samples
```

Comment:
You could get away with just declaring a `dict` here. The `defaultdict` only provides an advantage if you want to call `dict_of_lists[key].append()` without first calling `dict_of_lists[key] = []`.

Reply:
Good point 👍